src/dependency_injector/ext/starlette.py

```diff
@@ -0,0 +1,53 @@
+import sys
+from abc import ABCMeta, abstractmethod
+from typing import Any, Callable, Coroutine, Optional
+
+if sys.version_info >= (3, 11):  # pragma: no cover
+    from typing import Self
+else:  # pragma: no cover
+    from typing_extensions import Self
+
+from dependency_injector.containers import Container
+
+
+class Lifespan:
+    """A starlette lifespan handler performing container resource initialization and shutdown.
+
+    See https://www.starlette.io/lifespan/ for details.
+
+    Usage:
+
+    .. code-block:: python
+
+        from dependency_injector.containers import DeclarativeContainer
+        from dependency_injector.ext.starlette import Lifespan
+        from dependency_injector.providers import Factory, Self, Singleton
+        from starlette.applications import Starlette
+
+        class Container(DeclarativeContainer):
+            __self__ = Self()
+            lifespan = Singleton(Lifespan, __self__)
+            app = Factory(Starlette, lifespan=lifespan)
+
+    :param container: container instance
+    """
+
+    container: Container
+
+    def __init__(self, container: Container) -> None:
+        self.container = container
+
+    def __call__(self, app: Any) -> Self:
+        return self
+
+    async def __aenter__(self) -> None:
```

Comment:
Here you need to change the return type probably to `Optional[Awaitable]`

Reply:
1. This method will be inferred as `def __aenter__(self) -> Awaitable[None]:`
2. `-> Optional[Awaitable]` means function returns either `None` or something that you can `await` (e.g. signature of the `init_resources`). Given the pt.1, annotating it with your suggestion will result in `Awaitable[Optional[Awaitable]]`, which does not conforms to [StatelessLifespan](https://github.com/encode/starlette/blob/c2e3a39b09a613553ee03586589ed9cd0fbf07f3/starlette/types.py#L18) protocol.